rfcs/syntax-type-predicates.md

```diff
@@ -0,0 +1,155 @@
+
+Consider the following function which is called by a uncontrolled client (i.e., any arguments can actually be passed into it):
+
+```lua
+type Pet = {
+    name: string,
+    age: number,
+}
+
+local function processClientEvent(pet: unknown)
+    -- there is no way here to assert that pet is of type Pet
+end
+```
+
+## Design
+
+We propose a new syntax for return types in a function: `variable :: type`, where `variable` must match
+the name of a function parameter. In context:
+
+```lua
+type Pet = {
+    name: string,
+    age: number,
+}
+
+local function isPet(x: unknown): x :: Pet
+    return typeof(x) == "table" and typeof(x.name) == "string" and typeof(x.age) == "number"
+end
+```
+
+Function type syntax can follow a similar structure:
+
+```lua
+type isPet = (x: unknown) -> x :: Pet
+```
+
+We can then use this in practice:
+
+```lua
+local function processClientEvent(pet: unknown)
+    assert(isPet(pet), "invalid argument #1 to 'processClientEvent' (expected 'Pet')")
+    -- here we now know that pet is of type Pet
+    print(pet.name) -- no error
+    print(pet.age) -- no error
+    print(pet.size) -- TypeError
+end
+```
+
+User defined type predicates can also be used in type guards, and control flow:
+
+```lua
+local function processClientEvent(pet: unknown)
+    if isPet(pet) then
+        -- pet is "Pet"
+    end
+    -- pet is "unknown"
+end
+
+local function processClientEvent(pet: unknown)
+    if not isPet(pet) then return end
+    -- pet is "Pet"
+end
+```
+
+A type predicate function must return a boolean expression:
+
+```lua
+local function isPet(x: unknown): x :: Pet
+    return 5 -- TypeError: Type 'number' could not be converted into 'boolean'
+end
+```
+
+If the variable specified is not a parameter to the function, we will emit a type error:
+
+```lua
+local function isPet(pet: unknown): x :: Pet -- TypeError: Unknown parameter 'x'
+end
+```
+
+If the type we want to assert does not exist, we will emit a type error:
+
+```lua
+local function isDog(dog: unknown): dog :: Dog -- TypeError: Unknown type 'Dog'
+end
+```
+
+## Drawbacks
+
+One problem with user-defined type predicates, is that they do not completely solve the unsoundness issue, if the user
+incorrectly defines their type predicate function.
+
+For example:
+
+```lua
+type Pet = {
+    name: string,
+    age: number,
+    size: number,
+}
+
+local function isPet(x: unknown): x :: Pet
+    -- bug: we have not verified that x contains a property "size"
+    return typeof(x) == "table" and typeof(x.name) == "string" and typeof(x.age) == "number"
+end
+
+local function processClientEvent(pet: unknown)
+    assert(isPet(pet))
+    pet.size += 5 -- runtime error: attempt to perform arithmetic (add) on nil and number
+end
+
+processClientEvent({ name = "Cat", age = 5 })
+```
+
+## Alternatives
+
+We could follow the syntax of [TypeScript type predicates](https://www.typescriptlang.org/docs/handbook/advanced-types.html):
```

Comment:
Unfortunately not possible to do. TypeScript guys are lucky to have `{}` as a delimiter from the return type and the body.

```lua
function f(x: unknown): x is (
                          ---- Looks like a function call!
```